code/modules/mob/living/carbon/give.dm

```diff
@@ -1,52 +1,164 @@
+ */
+/obj/screen/alert/take_item
+	name = "Take Item"
+	desc = "someone wants to hand you an item!"
+	icon_state = "template"
+	timeout = 10 SECONDS
+	/// UID of the mob offering the receiver an item.
+	var/giver_UID
+	/// UID of the mob who has this alert.
+	var/receiver_UID
+	/// UID of the item being given.
+	var/item_UID
+
+
+/obj/screen/alert/take_item/Initialize(mapload, mob/giver, mob/receiver, obj/item/I)
+	. = ..()
+	desc = "[giver] wants to hand you \a [I]. Click here to accept it!"
+	giver_UID = giver.UID()
+	receiver_UID = receiver.UID()
+	item_UID = I.UID()
+	add_overlay(icon(I.icon, I.icon_state, SOUTH))
+	add_overlay("alert_flash")
+	ADD_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
+	RegisterSignal(I, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED), .proc/cancel_give)
+	RegisterSignal(giver, list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS), .proc/cancel_give)
+
+
+/obj/screen/alert/take_item/Destroy()
+	var/mob/living/giver = locateUID(giver_UID) // REMOVE_TRAIT needs a direct var reference.
+	REMOVE_TRAIT(giver, TRAIT_OFFERING_ITEM, TRAIT_GENERIC)
+	unregister_signals()
+	return ..()
+
+
+/obj/screen/alert/take_item/proc/unregister_signals()
+	UnregisterSignal(locateUID(giver_UID), list(SIGNAL_ADDTRAIT(TRAIT_HANDS_BLOCKED), COMSIG_CARBON_SWAP_HANDS))
+	UnregisterSignal(locateUID(item_UID), list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
 
-	if(target.incapacitated() || usr.incapacitated() || target.client == null)
+
+/obj/screen/alert/take_item/proc/cancel_give()
+	SIGNAL_HANDLER
+	var/mob/living/giver = locateUID(giver_UID)
+	to_chat(giver, "<span class='warning'>You need to keep the item in your active hand if you want to hand it to someone!</span>")
+	to_chat(locateUID(receiver_UID), "<span class='warning'>[giver] seems to have given up on giving you [locateUID(item_UID)].</span>")
+	qdel(src)
+
+
+/obj/screen/alert/take_item/Click(location, control, params)
+	var/mob/living/receiver = locateUID(receiver_UID)
+	var/obj/item/I = locateUID(item_UID)
+	if(HAS_TRAIT(receiver, TRAIT_HANDS_BLOCKED))
+		to_chat(receiver, "<span class='warning'>You need to have your hands free to accept [I]!</span>")
+		return
+	var/mob/living/giver = locateUID(giver_UID)
+	if(receiver.r_hand && receiver.l_hand)
+		to_chat(receiver, "<span class='warning'>Your hands are full!</span>")
 		return
+	if(!giver.Adjacent(receiver))
+		to_chat(receiver, "<span class='warning'>You need to stay in reaching distance of [giver] to take [I]!</span>")
+		return
+	if((I.flags & NODROP))
+		to_chat(giver, "<span class='warning'>[I] stays stuck to your hand when [receiver] tries to take it!</span>")
+		to_chat(receiver, "<span class='warning'>[I] stays stuck to [giver]'s hand when you try to take it!</span>")
+		return
+	// Give is successful at this point, unregister signals so the dropped signal doesn't fire and call `cancel_give()`.
+	unregister_signals()
+	giver.unEquip(I)
+	receiver.put_in_hands(I)
```

Comment:
does this work sanely if the receiver is lacking hand(s)?

Reply:
Yes. There are hand checks further in the call stack for this proc.